A near complete rewrite

jdenen opened the pull request:
- [x] Read a format from the format's module
- [x] `Geomancer.GeoJSON` shouldn't know about `Exshape` structs
- [x] `Geomancer` should delegate all conversion logic to format specific modules
- [x] Change paradigm from `FromFormatModule.to_format_function` to `ToFormatModule.from/1`
- [x] Use behaviour to enforce `from/1` and `read/1` functions
- [x] More unit tests

jdenen committed:
Use new read/1 and from/1 behaviour functions

- Make Geomancer a usable behaviour
- Implement read/1 and from/1 functions for GeoJson
- Implement read/1 and (unsupported) from/1 functions for Shapefile
- Use placebo to mock
- Run formatter

jdenen committed:
Move GeoJSON feature conversion logic to its own module

jdenen committed:
Add verify alias to mix

jdenen committed:
Add Geomancer.geo_json/1 delegate

- Remove outdated moduledoc to be fixed later

jdenen commented:
@paul-jacob @nathanielscottstevens @jakeprem @cameronmarsh

Y'all might be interested in reviewing

ManApart commented on lib/geomancer.ex:
```diff
@@ -1,21 +1,19 @@
-  @moduledoc """
-  A library to convert between various geospatial file formats
-  """
+  @moduledoc false
```
Why remove the moduledoc?

jdenen replied:
I'm going to do another pass at a later time to write moduledocs for everything. I didn't want to overlook this one that already has one, so I removed it. Kind of TODO'ing it, I guess.

ManApart commented on lib/geomancer/geo_json.ex:
```diff
@@ -15,9 +16,34 @@ defmodule Geomancer.GeoJson do
+        input_path
+        |> Geomancer.Shapefile.read()
+        |> new()
+        |> Jason.encode()
```
Seems weird to new up the struct and then immediately Jason encode it.

jdenen replied:
🤔

ManApart commented on lib/geomancer/geo_json/feature_set.ex:
```diff
@@ -0,0 +1,64 @@
+    values
+    |> Enum.map(&trim_dbf_value/1)
+    |> Enum.zip(keys)
+    |> Enum.map(fn {val, key} -> {key, val} end)
```
This made me do a double take. But on second pass it 'just makes sense'.

jdenen replied:
Thanks for pointing this one out. I've got an idea to clean it up.

ManApart commented on lib/geomancer/geo_json/feature_set.ex:
```diff
@@ -0,0 +1,64 @@
+defmodule Geomancer.GeoJson.FeatureSet do
```
With all of these maps and zips, this module displays an intimate knowledge of the structures of GeoJson and Shape files. I think a couple of 'why' comments would be super helpful to those who are reading this code in the future without as intimate a knowledge.

jdenen replied:
Yeah, definitely. Comments, moduledocs, and function docs are incoming.

NathanielScottStevens commented on lib/geomancer.ex:
```diff
@@ -1,21 +1,19 @@
+  @type input_path() :: String.t()
+  @type conversion() :: String.t()
+  @type reason() :: String.t()
+  @type geo_struct() :: struct | map
```
I find this type to be confusing. At the very least I would post_fix it with something besides struct since it can also be a map. I'm not sure that it makes sense to type something that can be either a shapefile struct or a geo_json decoded map. They seem too different though I understand you need something to give the callback.

NathanielScottStevens commented on lib/geomancer.ex:
```diff
@@ -1,21 +1,19 @@
-        {:error, "Unsupported format: #{ext}"}
+  defmacro __using__(_) do
+    quote do
+      @behaviour Geomancer
```
I would rather the `@behaviour` be added explicitly to geo_json and shapefile unless the using is going to contain more in the future.

jdenen replied:
I believe it is, but let me think through that.

NathanielScottStevens commented on lib/geomancer.ex:
```diff
@@ -1,21 +1,19 @@
-  @type geo_json :: String.t()
-  @type reason :: String.t()
+  @callback from(input_path) :: {:ok, conversion} | {:error, reason}
+  @callback read(input_path) :: {:ok, geo_struct} | {:error, reason}
```
I find `from` and `read` to be synonymous in this context. Could you perhaps rename `from` to something like `transform` or `translate`?

NathanielScottStevens replied:
Also, is the idea that this library will be handling more file formats in the future? If not, I'm skeptical of using behaviours here. (But I think it is)

jdenen replied:
They're now `read/1` and `convert/1`. Also, I do expect to have more file formats in the future.

NathanielScottStevens commented on lib/geomancer/geo_json.ex:
```diff
@@ -15,9 +17,88 @@ defmodule Geomancer.GeoJson do
-    bounding = [bbox.xmin, bbox.ymin, bbox.xmax, bbox.ymax]
-    %__MODULE__{features: features, name: name, bbox: bounding}
+  @spec new({:ok, struct()} | {:error, String.t()}) :: t() | {:error, String.t()}
+  def new({:ok, %Geomancer.Shapefile{} = shapefile}) do
```
I feel real weird about having an :ok tuple as an argument. I'd feel better about it if this were a private method. Otherwise I'd rather see it handled with a "railroad with" in the from but maybe I'm just being sorta pedantic.

jdenen replied:
🤔

NathanielScottStevens commented on lib/geomancer/geo_json.ex:
```diff
@@ -15,9 +17,88 @@ defmodule Geomancer.GeoJson do
+  @spec new({:ok, struct()} | {:error, String.t()}) :: t() | {:error, String.t()}
+  def new({:ok, %Geomancer.Shapefile{} = shapefile}) do
+    feature_set = features(shapefile.shp, shapefile.type, shapefile.dbf)
+    bbox = [shapefile.bbox.xmin, shapefile.bbox.ymin, shapefile.bbox.xmax, shapefile.bbox.ymax]
```
Pattern match out bbox?

jdenen replied:
Good call

NathanielScottStevens commented on lib/geomancer/geo_json.ex:
```diff
@@ -15,9 +17,88 @@ defmodule Geomancer.GeoJson do
+  @impl Geomancer
+  def from(input_path) do
+    case Path.extname(input_path) do
+      ".zip" ->
```
I'm not a fan of forcing the file to have a zip extension, though I can't think of another easy way of trying to enforce the correct file being passed in. The issue is cases like we have in reaper where the file is being downloaded as a `.shapefile` and needs to be renamed before being handed off to geomancer.

jdenen replied:
Why is it being downloaded as `.shapefile`? That's not valid at all. `.zip` is the proper extension for a Shapefile.

NathanielScottStevens replied:
I think they decided to give extensions that match the filetype which I'm not onboard with. However, I think their are other scenarios where this being a problem would be more valid (e.g. a new geo file type is created that also has .zip, or someone saves the file without an extension)

jdenen replied:
I'm going to ignore this in this PR and follow up with another.

jdenen committed:
Change from/1 to convert/1

jdenen committed:
Address PR feedback

- Make using macro more useful
- Make some public functions private
- Cleanup :ok vs :error tuple logic
- Other small tweaks

jdenen commented:
@NathanielScottStevens @ManApart Take another look and let me know what you think

jdenen committed:
Only feature_reducer/2 function heads are aware of source structs

NathanielScottStevens commented on lib/geomancer/geo_json/feature_set.ex:
```diff
@@ -47,18 +44,16 @@ defmodule Geomancer.GeoJson.FeatureSet do
-  defp trim_dbf_value(value), do: value
+  defp trim(value), do: value
+
+  defp parse_bbox(%{bbox: bbox}), do: bbox
```
Nice!

jdenen committed:
Move feature type logic to GeoJson.Feature module

- Add support for LineString feature type
